cloudevents/generic/options/grpc/agentoptions.go

```diff
@@ -0,0 +1,76 @@
+package grpc
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	cloudevents "github.com/cloudevents/sdk-go/v2"
+	cecontext "github.com/cloudevents/sdk-go/v2/context"
+
+	"open-cluster-management.io/api/cloudevents/generic/options"
+	"open-cluster-management.io/api/cloudevents/generic/options/grpc/protocol"
+	"open-cluster-management.io/api/cloudevents/generic/types"
+)
+
+type grpcAgentOptions struct {
+	GRPCOptions
+	errorChan   chan error // grpc client connection doesn't have error channel, it will handle reconnecting automatically
+	clusterName string
+}
+
+func NewAgentOptions(grpcOptions *GRPCOptions, clusterName, agentID string) *options.CloudEventsAgentOptions {
+	return &options.CloudEventsAgentOptions{
+		CloudEventsOptions: &grpcAgentOptions{
+			GRPCOptions: *grpcOptions,
+			errorChan:   make(chan error),
+			clusterName: clusterName,
+		},
+		AgentID:     agentID,
+		ClusterName: clusterName,
+	}
+}
+
+func (o *grpcAgentOptions) WithContext(ctx context.Context, evtCtx cloudevents.EventContext) (context.Context, error) {
+	eventType, err := types.ParseCloudEventsType(evtCtx.GetType())
```

Comment:
do we need this for grpc?

Reply:
We require this information to assign the topic for event publish, and the topic will be included in the gRPC publish request, see: https://github.com/open-cluster-management-io/api/blob/3012fe56a2f617cfa03e2f55f660c6f530e437d4/cloudevents/generic/options/grpc/protobuf/v1/cloudevent.proto#L68-L73